Refactor the workflow to process and validate submitted files in CWS

lerks opened the pull request:
Move the code out of the handler to a separate module. Split it into several functions, in order for it to be more modular, reusable and testable. Add extensive docstrings and unit tests.

This change makes the code longer but easier to understand as each function takes care of a single specific duty (before, they were all tangled steps). It also clarifies the semantics, where the intended behaviours are documented, exemplified and verified in docstrings and unit tests (before, some quirks emerged from interactions between far parts of the code and from a history of layered incomplete or incorrect changes). Moreover it reduces code duplication between submissions and user tests.

Some differences in semantics were introduced:
* before, submitting an archive was only allowed in language-agnostic settings; now it's always allowed (even though the interface hasn't yet been updated to reflect this).
* before, the language had always to be explicitly provided, now it will be guessed if necessary (from all allowed languages or from a user-provided set of candidates).
* before, when submitting an incremental user test on a task that allowed it only the missing source files would be fetched from the previous test, not the missing input or managers, this is now fixed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cms-dev/cms/945)
<!-- Reviewable:end -->

stefano-maggiolo commented:
Sorry, only partial review for the moment. Missing all tests and looked at the new functions only up to the last comment I placed.

---

Reviewed 2 of 6 files at r1.
Review status: 2 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[cms/grading/language.py, line 156 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAhznIM9YEk2gDj09Nz:-LAhznIM9YEk2gDj09O-:b-b4da0x) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/grading/language.py#L156)):*
> ```Python
>     # keys. Since they have no state (they are just collections of
>     # constants and static methods) and are designed to be used as
>     # singletons, it's very easy to make them hashable.
> ```

Do you want to actually make them singleton? There is an example in functionaltestframework.py

---

*[cms/server/contest/submission.py, line 197 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj-usx1T5OpRlPZYbx:-LAj-usx1T5OpRlPZYby:b-eevrq) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L197)):*
> ```Python
> # is the name of the form field (in our case it's the filename-with-%l)
> # and filename is the name of the file had on the user's system.
> ReceivedFile = namedtuple("ReceivedFile", ["codename", "filename", "content"])
> ```

Types? I assume str str bytes, but maybe specify in the comment.

---

*[cms/server/contest/submission.py, line 215 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj06_x1-xpGIGt-lHk:-LAj06_x1-xpGIGt-lHl:bwyfe1a) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L215)):*
> ```Python
>     will be ignored and the files will be returned with their basename.
> 
>     data (bytes): the encoded contents of the archive.
> ```

"encoded"?

---

*[cms/server/contest/submission.py, line 217 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj0FdD7ugJlnWYVtk-:-LAj0FdD7ugJlnWYVtk0:b-cgwxpq) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L217)):*
> ```Python
>     data (bytes): the encoded contents of the archive.
> 
>     return ([ReceivedFile]): the files contained in the archive.
> ```

What's codename here?

---

*[cms/server/contest/submission.py, line 267 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj0gim8dqIdfg6Rx8n:-LAj0gim8dqIdfg6Rx8o:bfbike2) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L267)):*
> ```Python
>         return extract_files_from_archive(tornado_files["submission"][0].body)
> 
>     else:
> ```

No need for else:

---

*[cms/server/contest/submission.py, line 276 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj0kbc59PxBZ-vN_KR:-LAj0kbc59PxBZ-vN_KS:b-shnpzi) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L276)):*
> ```Python
> 
> 
> class InvalidFiles(Exception):
> ```

Are these exception tightly coupled to the functions? I think it would be clearer if they were all on top. Don't feel strongly about it though.

---

*[cms/server/contest/submission.py, line 291 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj2A_w1hyCtiktQrzE:-LAj2A_w1hyCtiktQrzF:b-z39ozs) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L291)):*
> ```Python
>     - the codename isn't given and the filename matches element of the
>       submission format when we replace its trailing ".%l" (if any) with
>       one of the source extensions of the language.
> ```

Nit: the order here is the opposite of the order in the code.

---

*[cms/server/contest/submission.py, line 315 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj1x-t2Y0Kcjbjn2Cy:-LAj1x-t2Y0Kcjbjn2Cz:b2v9njl) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L315)):*
> ```Python
>                 if language is None:
>                     raise ValueError(
>                         "language not given when submission format requires it")
> ```

Why not capitalized?

---

*[cms/server/contest/submission.py, line 316 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj7Owd33sPzNrVuaQz:-LAj7Owd33sPzNrVuaR-:bt8ym5w) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L316)):*
> ```Python
>                     raise ValueError(
>                         "language not given when submission format requires it")
>                 base = element.rpartition(".")[0]
> ```

os.path.splitext?

---

*[cms/server/contest/submission.py, line 329 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj2aKM4eLzkCBNRae3:-LAj2aKM4eLzkCBNRae4:b7xpyfr) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L329)):*
> ```Python
>                 "language not given when submission format requires it")
>         elif any(filename.endswith(ext) for ext in language.source_extensions):
>             candidate_elements.add(codename)
> ```

Isn't equivalent and simpler?
```python
for element in submission_format:
    if codename is not None and codename != element:
        # Early exit if provided codename doesn't match this element.
        continue
    if element.endswith(".%l"):
        if language is None:
            raise ValueError(
                "language not given when submission format requires it")
        filename_ext_matches = filename is not None and any(
                    filename.endswith(ext) for ext in language.source_extensions)
        if codename is not None:
            # When codename is not None, matches if filename is None, or its
            # extension matches the language.
            if filename is None or filename_ext_matches:
                candidate_elements.add(codename)
        else:
            # When codename is None, matches if filename matches both in
            # basename and language extension.
            element_base = element.rpartition(".")[0]
            filename_base = filename.rpartition(".")[0]
            if filename_base = element_base and filename_ext_matches:
                candidate_elements.add(codename)
    else:
        if codename is not None or filename == element:
            candidate_elements.add(codename)
```

or even
```python
for element in submission_format:
    if codename is not None:
        if _match_file_with_codename(codename, filename, language, element):
            candidate_elements.add(codename)
    else:
        if _match_file_without_codename(filename, language, element):
            candidate_elements.add(codename)

[...]

def _match_file_with_codename(codename, filename, language, element):
    if codename != element:
        return False
    if not element.endswith(".%l"):
        return True
    return any(filename.endswith(ext) for ext in language.source_extensions)

def _match_file_without_codename(filename, language, element):
    if not element.endswith(".%l"):
        return filename == element
    element_base = os.path.splitext(element)[0]
    filename_base, filename_ext = os.path.splitext(filename)
    return filename_base = element_base and any(
        filename_ext == ext for ext in language.source_extensions)
```

---

*[cms/server/contest/submission.py, line 340 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAjBfOR9lePjy66DedB:-LAjBfOR9lePjy66DedC:b-289ysk) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L340)):*
> ```Python
>             % (codename, filename))
> 
>     return next(iter(candidate_elements))
> ```

candidate_elements.pop()

---

*[cms/server/contest/submission.py, line 367 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAjBv3HEeV9TUCsxHSV:-LAjBv3HEeV9TUCsxHSW:b2ni2zd) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L367)):*
> ```Python
>         if codename in files:
>             raise InvalidFiles(
>                 "fwo files match the same element %r of the submission format"
> ```

fwo

---

*[cms/server/contest/submission.py, line 405 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAjCKLODv8ir29VNT3D:-LAjCKLODv8ir29VNT3E:b3pbpkk) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L405)):*
> ```Python
> 
>     given_files ([ReceivedFile]): the submitted files.
>     given_language_names ({str}|None): a set of languages, for example
> ```

why are you not requiring a single language or None? can't you terminate the request if you get more than one from the user, which is obviously wrong?

---

*[cms/server/contest/handlers/tasksubmission.py, line 139 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAi3MWGAx-ol6ilpF8k:-LAi3MWGAx-ol6ilpF8l:bsjiaur) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/handlers/tasksubmission.py#L139)):*
> ```Python
> 
>         # Required files from the user.
>         required = set(task.submission_format)
> ```

required_files? Reading later I thought this was a boolean

---

*[cms/server/contest/handlers/tasksubmission.py, line 152 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAi3CEH4hDz8LjriogC:-LAi3CEH4hDz8LjriogD:b5xzuwn) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/handlers/tasksubmission.py#L152)):*
> ```Python
>             given_languages = {self.get_argument("language")}
>         except tornado.web.MissingArgumentError:
>             given_languages = None
> ```

Comment about what this means (I guess either the submission didn't need a language, or autodetect will happen?)

---

*[cms/server/contest/handlers/tasksubmission.py, line 164 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAi3vgo1YU-2ESkcQBc:-LAi3vgo1YU-2ESkcQBd:b-3ywp14) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/handlers/tasksubmission.py#L164)):*
> ```Python
> 
>         file_digests = dict()
>         missing = required.difference(iterkeys(files))
> ```

Similarly, missing_files?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/945#-:-LAjCWha3jF_EojrdIaT:bqf3tyx)*
<!-- Sent from Reviewable.io -->

lerks commented:
Review status: 2 of 6 files reviewed at latest revision, 17 unresolved discussions.

---

*[cms/grading/language.py, line 156 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAhznIM9YEk2gDj09Nz:-LAkx_LdBOFyCKsQ4Y1o:b-9zgrye) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/grading/language.py#L156)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Do you want to actually make them singleton? There is an example in functionaltestframework.py
</blockquote></details>

Nah, they already essentially are (as long as they are obtained through `get_language`)

---

*[cms/server/contest/submission.py, line 197 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj-usx1T5OpRlPZYbx:-LAkuG6Z-Pr0omptN9rL:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L197)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Types? I assume str str bytes, but maybe specify in the comment.
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 215 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj06_x1-xpGIGt-lHk:-LAkuHAq4RoSGxyhQrlX:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L215)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

"encoded"?
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 217 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj0FdD7ugJlnWYVtk-:-LAkuHcwBwbrBWc6R6Bc:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L217)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

What's codename here?
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 267 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj0gim8dqIdfg6Rx8n:-LAkuIIKC2hkiflNreJX:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L267)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

No need for else:
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 276 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj0kbc59PxBZ-vN_KR:-LAkunZy1G7kkV8dGhDd:b-qmpxzp) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L276)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Are these exception tightly coupled to the functions? I think it would be clearer if they were all on top. Don't feel strongly about it though.
</blockquote></details>

Well, sort of: each exception is there to be used by the 1-2 following functions to signal an insuccess.

---

*[cms/server/contest/submission.py, line 291 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj2A_w1hyCtiktQrzE:-LAkuJHB9BevVa1gm2Hn:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L291)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Nit: the order here is the opposite of the order in the code.
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 315 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj1x-t2Y0Kcjbjn2Cy:-LAkv1eg80nxD5qbth5t:b-q70zem) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L315)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Why not capitalized?
</blockquote></details>

I don't think I've ever been consistent on this. For example the exceptions raised by the other functions that already were in this file also have lowercase messages. I have no problem to change them all: should I?

---

*[cms/server/contest/submission.py, line 316 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj7Owd33sPzNrVuaQz:-LAkuJwC3AUEnHcOGZVK:bjz44sw) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L316)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

os.path.splitext?
</blockquote></details>

Done.

Also, this can only be safely done on the codename/element, not on the filename: I added a test for that.

---

*[cms/server/contest/submission.py, line 340 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAjBfOR9lePjy66DedB:-LAkuSTyEmC7t0dJDwWO:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L340)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

candidate_elements.pop()
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 367 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAjBv3HEeV9TUCsxHSV:-LAkuT8A5oP9rM-ZmMYH:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L367)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

fwo
</blockquote></details>

Done.

---

*[cms/server/contest/handlers/tasksubmission.py, line 139 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAi3MWGAx-ol6ilpF8k:-LAkuV9S1g4JARWPxriK:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/handlers/tasksubmission.py#L139)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

required_files? Reading later I thought this was a boolean
</blockquote></details>

Done.

---

*[cms/server/contest/handlers/tasksubmission.py, line 164 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAi3vgo1YU-2ESkcQBc:-LAkuVfrBuYXiVxFNyxE:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/handlers/tasksubmission.py#L164)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Similarly, missing_files?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/945)*
<!-- Sent from Reviewable.io -->

stefano-maggiolo commented:
Points to debate:
- is there a way to split submission.py meaningfully?
- why don't allow user-provided languages to be invalid?
- more in general, I'm more for allowing a single user-provided language - complexity is already very high and I don't see a clear use case for multiple. Actually, there might be some benefit for code complexity to make the parsing a bit stricter - for example looking first at whether the SF is languageless, and in that case not allowing user-provided languages

---

Reviewed 2 of 2 files at r4.
Review status: 4 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[cms/grading/language.py, line 156 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAhznIM9YEk2gDj09Nz:-LAlZ6EWEWnr_rFr69FK:b32w1t) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/grading/language.py#L156)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Nah, they already essentially are (as long as they are obtained through `get_language`)
</blockquote></details>

Ok

---

*[cms/server/contest/submission.py, line 315 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAj1x-t2Y0Kcjbjn2Cy:-LAlZYs62VXga13tQaqr:bwvdvc5) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L315)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

I don't think I've ever been consistent on this. For example the exceptions raised by the other functions that already were in this file also have lowercase messages. I have no problem to change them all: should I?
</blockquote></details>

Let's do it later, in case.

---

*[cms/server/contest/submission.py, line 416 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlhs3H7OA04jFoozcB:-LAlhs3IB68gOYELrg-0:bzdzh52) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L416)):*
> ```Python
> 
>     return ({str: bytes}, Language|None): the mapping from codenames to
>         content, and the language of the submission.
> ```

Explain None

---

*[cms/server/contest/submission.py, line 434 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAliUfCB_Xo_-e-jzbw:-LAliUfCB_Xo_-e-jzbx:bcgkghh) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L434)):*
> ```Python
>                     pass
>                 else:
>                     candidate_languages.add(language)
> ```

spare two lines and do try: `c_l.add(g_l(l_n))`? (also later)

---

*[cms/server/contest/submission.py, line 436 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlieGS4dIa3dkVnTst:-LAlieGS4dIa3dkVnTsu:b-kyy94x) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L436)):*
> ```Python
>                     candidate_languages.add(language)
>         else:
>             candidate_languages = set(LANGUAGES)
> ```

Nit: quick, positive case first?

---

*[cms/server/contest/submission.py, line 454 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAljChLBDipkBy4tGWQ:-LAljChLBDipkBy4tGWR:ba5uc4z) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L454)):*
> ```Python
>                 .difference(allowed_language_names)
>             if len(forbidden_language_names) > 0:
>                 raise InvalidFilesOrLanguages(
> ```

Shouldn't you take the intersection here, and raising only if it is empty?

---

*[cms/server/contest/submission.py, line 459 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlkKnh00zMQX4S4e1O:-LAlkKnh00zMQX4S4e1P:bg9fe9i) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L459)):*
> ```Python
> 
>     if not any(element.endswith(".%l") for element in submission_format):
>         candidate_languages = {None}
> ```

So we worked so hard and now we throw everything away? Maybe put the code above in a function and call it only if necessary? Or do you want to retain the exceptions in case of invalid (but unused) input?

---

*[cms/server/contest/submission.py, line 480 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlkyixAgVR3b3n1Nrf:-LAlkyixAgVR3b3n1Nrg:bo8qzp8) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L480)):*
> ```Python
>             % set(iterkeys(matched_files_by_language)))
> 
>     language, files = next(iteritems(matched_files_by_language))
> ```

.popitem()

---

*[cms/server/contest/submission.py, line 512 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAllYPv-MakwxKcbh04:-LAllYPw1kdK7lDPoCxc:bq17vry) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L512)):*
> ```Python
>     """
>     # FIXME Instead of taking the latest submission *if* it is of the
>     # right language, take the latest *among* those of that language.
> ```

Before we were taking the latest submission regardless of the language, right? (Anyway, this only applies to OO tasks, where the language is None, so this is not too important).

---

*[cms/server/contest/submission.py, line 526 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlqRxYCSkArAU2mB3p:-LAlqRxYCSkArAU2mB3q:b75ibqi) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L526)):*
> ```Python
>         elif cls is UserTest:
>             if codename == "input":
>                 digests["input"] = latest_submission.input
> ```

IIUC you're merging two namespaces here, but there is nothing enforcing no conflicts will arise, no?

---

*[cms/server/contest/submission.py, line 533 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlls35AMoUMM6e_Wg3:-LAlls35AMoUMM6e_Wg4:b4bwij0) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L533)):*
> ```Python
>                                          "format requires it")
>                     filename = \
>                         codename.rpartition(".")[0] + language.source_extension
> ```

splitext

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 451 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlr54tDPIu5H_tyIhD:-LAlr54tDPIu5H_tyIhE:b-c44w5e) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L451)):*
> ```Python
>                 f.addfile(tarinfo, fileobj)
>         six.assertCountEqual(
>             self, extract_files_from_archive(archive_data.getvalue()), files)
> ```

Nit: use the same structure for test_zip and test_tar_gz?

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 459 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlrDPn6TWItyfH9Gst:-LAlrDPn6TWItyfH9Gsu:b-mf42fl) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L459)):*
> ```Python
>     # Make sure we ignore the directory structure and only use the
>     # trailing component (i.e., the basename) in the return value, even
>     # if it leads to duplicated filenames.
> ```

Comments about the test inside the test (several times)

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 503 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmVgvc2aGNhvRgoypi:-LAmVgvc2aGNhvRgoypj:bsj084q) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L503)):*
> ```Python
>         # The extracted file is named like the temporary file where the
>         # archive's contents were copied to, plus a trailing tilde.
>         six.assertRegex(self, f.filename, "tmp[a-z0-9_]+~")
> ```

Why do you want to make sure the name is this?

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 517 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmW1B9Eek3TCDXHnky:-LAmW1B9Eek3TCDXHnkz:bdi34bg) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L517)):*
> ```Python
>     # This should check that the extracted files cannot "escape" from
>     # the temporary directory where they are being extracted to.
>     def test_weird_filenames(self):
> ```

test_double_dot?

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 531 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmW9fz23H4ACrZyYxm:-LAmW9fz23H4ACrZyYxn:b-umkmmu) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L531)):*
> ```Python
>     # zipfile, tarfile, etc. directly but it would be too burdensome to
>     # support the same amount of archive types as patool does.
>     def test_conflicting_filenames(self):
> ```

This also seems a bit of a "who cares?"-test. If it raises, ok, if it doesn't, something will come out of it, probably wrong, and the user that made such a bad zip archive will get something random.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 772 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmcHCj5z-i0WjgLSQp:-LAmcHCj5z-i0WjgLSQq:b-vnjype) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L772)):*
> ```Python
>         # function doesn't split the extension on the filename.
>         files, language = match_files_and_languages(
>             [ReceivedFile(None, "foolib.pas", FOO_CONTENT)],
> ```

Also assertRaises for filename=foo.lib.pas?

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 808 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmcZEADtUVp43y1UOO:-LAmcZEADtUVp43y1UOP:b-2xkg5l) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L808)):*
> ```Python
> 
>         # This brings in some weird side-effects, for example matching
>         # an unexpected language as it's the only one left.
> ```

I don't get this, shouldn't this say "I was unable to find which language this submission is in"?

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 867 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmd0MtAx_Gqf6M_Cst:-LAmd0MtAx_Gqf6M_Csu:b-4fh6yf) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L867)):*
> ```Python
> 
>         # It fails also if any of the given languages is forbidden, even
>         # if some other one is given, allowed and matches.
> ```

Yeah, as above, I don't get why we raise in this case.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 952 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmdaayAViMiKbVZyss:-LAmdaayAViMiKbVZyst:b-xc8p9n) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L952)):*
> ```Python
>             match_files_and_languages(
>                 [ReceivedFile(None, None, FOO_CONTENT)],
>                 {"C"}, {"foo.%l"}, None)
> ```

This is true for languageless SFEs too, right? 

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 958 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmdyHZ4yYAqaBUkA-D:-LAmdyHZ4yYAqaBUkA-E:bbxpo5y) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L958)):*
> ```Python
> 
>         # Passing a language that doesn't exist means the contestant
>         # doesn't know what they are doing: we're not following through.
> ```

Again I'm for allowing this.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 1043 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmeUN88SDbwbEhS0fB:-LAmeUN88SDbwbEhS0fC:bpmt2or) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L1043)):*
> ```Python
> 
>         # If there are no files this could be made to work. However we
>         # decided that this means the contestant is very confused and
> ```

More like the contest is very confused.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 1181 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmg3eW4FHVv5IH5iKY:-LAmg3eW4FHVv5IH5iKZ:b-fg5dbi) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L1181)):*
> ```Python
>         self.insert_user_test("C", input_digest,
>                               {"foo.%l": foo_digest, "baz.txt": baz_digest},
>                               {"spam.c": spam_digest, "eggs.h": eggs_digest})
> ```

What if there is an "input" codename here?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/945#-:-LAmetIi3JGU6F-sv8fz:b-87mx23)*
<!-- Sent from Reviewable.io -->

lerks commented:
Review status: 2 of 6 files reviewed at latest revision, 25 unresolved discussions.

---

*[cms/server/contest/submission.py, line 405 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAjCKLODv8ir29VNT3D:-LAmx_W19aBj-Psl7iPi:b499q7j) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L405)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

why are you not requiring a single language or None? can't you terminate the request if you get more than one from the user, which is obviously wrong?
</blockquote></details>

Changed

---

*[cms/server/contest/submission.py, line 434 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAliUfCB_Xo_-e-jzbw:-LAmxWOl4Vo6TJf_Z4ej:b499q7j) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L434)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

spare two lines and do try: `c_l.add(g_l(l_n))`? (also later)
</blockquote></details>

Changed

---

*[cms/server/contest/submission.py, line 436 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlieGS4dIa3dkVnTst:-LAmxVNc6X35AoCi-gXc:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L436)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Nit: quick, positive case first?
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 454 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAljChLBDipkBy4tGWQ:-LAmxGYlAPV5_HkcPR2h:b499q7j) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L454)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Shouldn't you take the intersection here, and raising only if it is empty?
</blockquote></details>

Changed

---

*[cms/server/contest/submission.py, line 459 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlkKnh00zMQX4S4e1O:-LAmxEwbE0Y-C8sSRlgx:b499q7j) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L459)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

So we worked so hard and now we throw everything away? Maybe put the code above in a function and call it only if necessary? Or do you want to retain the exceptions in case of invalid (but unused) input?
</blockquote></details>

Changed

---

*[cms/server/contest/submission.py, line 512 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAllYPv-MakwxKcbh04:-LAmx3DHEd2b6luUUdVm:ba5enev) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L512)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Before we were taking the latest submission regardless of the language, right? (Anyway, this only applies to OO tasks, where the language is None, so this is not too important).
</blockquote></details>

Yes, I didn't change the behavior, just made it clearer and documented it. It might be cool if it worked outside of OO tasks too, though, don't you think?

---

*[cms/server/contest/submission.py, line 526 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlqRxYCSkArAU2mB3p:-LAmwn7V3h1Mupej4qiw:b-4lizzh) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L526)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

IIUC you're merging two namespaces here, but there is nothing enforcing no conflicts will arise, no?
</blockquote></details>

A (quite implicit?) assumption about user tests is that the submission format, the task type's get_user_managers and {input} are disjoint sets. Not sure what would happen when that is not the case...

BTW, it was like this before this commit

---

*[cms/server/contest/handlers/tasksubmission.py, line 152 at r1](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAi3CEH4hDz8LjriogC:-LAmwknRDGF0lMJqk3Eb:bb6wk04) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/handlers/tasksubmission.py#L152)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Comment about what this means (I guess either the submission didn't need a language, or autodetect will happen?)
</blockquote></details>

Gone

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 808 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmcZEADtUVp43y1UOO:-LAmwF8_23o-DXppcqiY:b5ktbgc) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L808)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

I don't get this, shouldn't this say "I was unable to find which language this submission is in"?
</blockquote></details>

Clarified in private

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 867 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmd0MtAx_Gqf6M_Cst:-LAmwLRcE8JJZ_-ILxV0:b3pncvg) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L867)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Yeah, as above, I don't get why we raise in this case.
</blockquote></details>

Test is gone

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 958 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmdyHZ4yYAqaBUkA-D:-LAmwQCGACeWpB7KIfiW:bhz7wtg) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L958)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Again I'm for allowing this.
</blockquote></details>

Interface changed, this should now make more sense

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 1043 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmeUN88SDbwbEhS0fB:-LAmwXDQBn-E2YouU8q9:bme12k6) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L1043)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

More like the contest is very confused.
</blockquote></details>

Do you want me to take any action here?

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 1181 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmg3eW4FHVv5IH5iKY:-LAmwdLA2B4AUHX95pdv:b-y6raml) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L1181)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

What if there is an "input" codename here?
</blockquote></details>

Does the above answer this?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/945)*
<!-- Sent from Reviewable.io -->

stefano-maggiolo commented:
Reviewed 3 of 4 files at r12, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[cms/server/contest/submission.py, line 512 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAllYPv-MakwxKcbh04:-LAnQpDl5JoPDMk6SObL:b-80tlqk) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L512)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Yes, I didn't change the behavior, just made it clearer and documented it. It might be cool if it worked outside of OO tasks too, though, don't you think?
</blockquote></details>

Sure.

---

*[cms/server/contest/submission.py, line 526 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlqRxYCSkArAU2mB3p:-LAnQtzP8q_-vAGRRC0b:b2nhll9) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L526)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

A (quite implicit?) assumption about user tests is that the submission format, the task type's get_user_managers and {input} are disjoint sets. Not sure what would happen when that is not the case...

BTW, it was like this before this commit
</blockquote></details>

Ok.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 808 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmcZEADtUVp43y1UOO:-LAnRJpS53IhHrc_Wlh4:bcb9ufl) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L808)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Clarified in private
</blockquote></details>

The clarification was: if language was C, then the passed foo.c is ambiguous (could match both foo.c and foo.%l), so C cannot be a match.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 1043 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmeUN88SDbwbEhS0fB:-LAnRgCF9QErrWoD3vIK:b-rhenor) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L1043)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Do you want me to take any action here?
</blockquote></details>

Change the comment - I don't see how an empty SF could be the contestant's fault

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 1181 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmg3eW4FHVv5IH5iKY:-LAnRmNv8oY-4OWI0UN7:b-h453e4) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L1181)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Does the above answer this?
</blockquote></details>

Ok, maybe put a clarification comment in the main code saying that we know this and judge it to be a configuration error in case it happens.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/945)*
<!-- Sent from Reviewable.io -->

lerks commented:
I've addressed all the smaller comments (i.e., all but the one where you rewrite my code :P ). Please take a look at these and then I'll fix the last one too.

---

Review status: 2 of 6 files reviewed at latest revision, 13 unresolved discussions.

---

*[cms/server/contest/submission.py, line 416 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlhs3H7OA04jFoozcB:-LAorBIV9ND6t6z3fm16:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L416)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Explain None
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 480 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlkyixAgVR3b3n1Nrf:-LAorRwM3vsC7Pw7o-10:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L480)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

.popitem()
</blockquote></details>

Done.

---

*[cms/server/contest/submission.py, line 533 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlls35AMoUMM6e_Wg3:-LAori3Q6UqPSluw_PNJ:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cms/server/contest/submission.py#L533)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

splitext
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 451 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlr54tDPIu5H_tyIhD:-LAps_yiCNwLc98oz4bw:b-ln40su) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L451)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Nit: use the same structure for test_zip and test_tar_gz?
</blockquote></details>

Done, though maybe not in the way you wanted ;)

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 459 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAlrDPn6TWItyfH9Gst:-LAossj5E09fj3Xx5GMX:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L459)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Comments about the test inside the test (several times)
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 503 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmVgvc2aGNhvRgoypi:-LApol3z3Vzx2pXUjnsE:b-10bxk6) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L503)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Why do you want to make sure the name is this?
</blockquote></details>

In order to substantiate the claim I am making in the comment.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 517 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmW1B9Eek3TCDXHnky:-LApoycHAA2oBvCbMZ8M:b-bjpj7a) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L517)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

test_double_dot?
</blockquote></details>

Better?

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 531 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmW9fz23H4ACrZyYxm:-LApsAABCip-qw6sVUE7:b1cupj5) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L531)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

This also seems a bit of a "who cares?"-test. If it raises, ok, if it doesn't, something will come out of it, probably wrong, and the user that made such a bad zip archive will get something random.
</blockquote></details>

Added a comment explaining the "spirit" of these tests.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 772 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmcHCj5z-i0WjgLSQp:-LAotT8N-CKluiiIJZj4:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L772)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Also assertRaises for filename=foo.lib.pas?
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 808 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmcZEADtUVp43y1UOO:-LAovaVs0AtJQVkK4yAt:b35wl8g) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L808)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

The clarification was: if language was C, then the passed foo.c is ambiguous (could match both foo.c and foo.%l), so C cannot be a match.
</blockquote></details>

Expanded the comment.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 952 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmdaayAViMiKbVZyss:-LAovruk-ojP5qXdszD9:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L952)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

This is true for languageless SFEs too, right? 
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 1043 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmeUN88SDbwbEhS0fB:-LAow90HB6bcbQ_kXs3Y:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L1043)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Change the comment - I don't see how an empty SF could be the contestant's fault
</blockquote></details>

Done.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 1181 at r11](https://beta.reviewable.io/reviews/cms-dev/cms/945#-LAmg3eW4FHVv5IH5iKY:-LAox6kf9cEe0ayBd8_p:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L1181)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Ok, maybe put a clarification comment in the main code saying that we know this and judge it to be a configuration error in case it happens.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cms-dev/cms/945#-:-LApt1yHCYkVF8TDFj5X:b-e0e9if)*
<!-- Sent from Reviewable.io -->

stefano-maggiolo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> good job on the complexity! :) 

---

Reviewed 2 of 4 files at r15, 1 of 1 files at r17, 1 of 1 files at r18.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cms/server/contest/submission.py, line 329 at r1](https://reviewable.io/reviews/cms-dev/cms/945#-LAj2aKM4eLzkCBNRae3:-LAqilmT7uxjI3opStQH:b-qn5lay) ([raw file](https://github.com/cms-dev/cms/blob/b6a7c9bedeed6009a7d8d785f7233af07c9ff046/cms/server/contest/submission.py#L329)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Isn't equivalent and simpler?
```python
for element in submission_format:
    if codename is not None and codename != element:
        # Early exit if provided codename doesn't match this element.
        continue
    if element.endswith(".%l"):
        if language is None:
            raise ValueError(
                "language not given when submission format requires it")
        filename_ext_matches = filename is not None and any(
                    filename.endswith(ext) for ext in language.source_extensions)
        if codename is not None:
            # When codename is not None, matches if filename is None, or its
            # extension matches the language.
            if filename is None or filename_ext_matches:
                candidate_elements.add(codename)
        else:
            # When codename is None, matches if filename matches both in
            # basename and language extension.
            element_base = element.rpartition(".")[0]
            filename_base = filename.rpartition(".")[0]
            if filename_base = element_base and filename_ext_matches:
                candidate_elements.add(codename)
    else:
        if codename is not None or filename == element:
            candidate_elements.add(codename)
```

or even
```python
for element in submission_format:
    if codename is not None:
        if _match_file_with_codename(codename, filename, language, element):
            candidate_elements.add(codename)
    else:
        if _match_file_without_codename(filename, language, element):
            candidate_elements.add(codename)

[...]

def _match_file_with_codename(codename, filename, language, element):
    if codename != element:
        return False
    if not element.endswith(".%l"):
        return True
    return any(filename.endswith(ext) for ext in language.source_extensions)

def _match_file_without_codename(filename, language, element):
    if not element.endswith(".%l"):
        return filename == element
    element_base = os.path.splitext(element)[0]
    filename_base, filename_ext = os.path.splitext(filename)
    return filename_base = element_base and any(
        filename_ext == ext for ext in language.source_extensions)
```
</blockquote></details>

Perfect.

---

*[cmstestsuite/unit_tests/server/contest/submission_test.py, line 531 at r11](https://reviewable.io/reviews/cms-dev/cms/945#-LAmW9fz23H4ACrZyYxm:-LAqkkMP5SYPpax5ZRdL:b-4yo44e) ([raw file](https://github.com/cms-dev/cms/blob/2d582db894d32435d293bac754e6068f79538763/cmstestsuite/unit_tests/server/contest/submission_test.py#L531)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Added a comment explaining the "spirit" of these tests.
</blockquote></details>

I'm still not sure why we have them, but if you care ok with the explanation.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cms-dev/cms/945#-:-LAqlOyv8mvG8wPzgvEY:bt87523)*
<!-- Sent from Reviewable.io -->
